R/view.R

```diff
@@ -24,28 +24,55 @@
 view <- function(x, title = NULL, ..., n = NULL) {
   check_dots_empty()
 
-  if (!interactive()) return(invisible(x))
+  if (!rlang::is_interactive()) {
+    return(invisible(x))
+  }
+
+  # The user's expression and the environment to re-evaluate it in
+  quo <- enquo0(x)
+  expr <- quo_get_expr(quo)
+  env <- quo_get_env(quo)
 
   if (is.null(title)) {
-    title <- expr_deparse(substitute(x), width = Inf)
+    title <- as_label(expr)
   }
 
+  # Retrieve the `View()` function, which includes the special
+  # hooks created by RStudio or Positron
+  fn <- get("View", envir = as.environment("package:utils"))
+
   if (!is.data.frame(x)) {
-    if (is.null(n)) {
-      n <- get_tibble_option_view_max()
-    }
-    x <- head(x, n + 1)
-    x <- as.data.frame(x)
-    if (nrow(x) > n) {
-      message("Showing the first ", n, " rows.")
-      x <- head(x, n)
-    }
+    return(view_with_coercion(x, n, title, fn))
+  }
+
+  # Make a `View()` call that we evaluate in the parent frame,
+  # as if the user called `View()` directly rather than `view()`.
+  # If `expr` directly references a data frame in the parent frame, this
+  # allows RStudio and Positron to "track" that original object
+  # for live updates in the data viewer.
+  inject((!!fn)(!!expr, !!title), env = env)
+
+  invisible(x)
+}
+
+view_with_coercion <- function(x, n, title, fn) {
+  if (is.null(n)) {
+    n <- get_tibble_option_view_max()
+  }
+
+  x <- head(x, n + 1)
+  x <- as.data.frame(x)
+
+  if (nrow(x) > n) {
+    message("Showing the first ", n, " rows.")
+    x <- head(x, n)
```

Comment:
This seems slower than it needs to be?

```suggestion
    x <- x[seq_len(n), , drop = FALSE]
```

I wonder if vctrs has a helper that creates an ALTREP powered subset?

Reply:
`vec_slice(x, seq_len(n))` would be the equivalent I guess

No ALTREP involved there though. We'd need ALTREP contiguous views into `x` to make that work. But that's not unheard of https://github.com/r-lib/rlang/pull/1725